capstone/src/test/java/com/google/sps/servlets/TagsTFIDFServletTest.java

```diff
@@ -0,0 +1,157 @@
+
+/**
+ * Unit tests for {@link TagsTFIDFServlet}.
+ */
+ @RunWith(JUnit4.class)
+public class TagsTFIDFServletTest {
+  private final long GROUP_ID_1 = 1;
+  private final long GROUP_ID_2 = 2;
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+              new LocalDatastoreServiceTestConfig()
+                  .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+              new LocalUserServiceTestConfig())
+          .setEnvIsLoggedIn(true);
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private LinkedHashMap<Long, LinkedHashMap<String, Integer>> groupMap;
+  private TagsTFIDFServlet tagsTFIDFServlet;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    populateDatabase(datastore);
+    setUpGroupMap();
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    tagsTFIDFServlet = new TagsTFIDFServlet();
+  }
+
+  @After
+  public void tearDown() {
+    // tear down work
+  }
+
+  @Test
+  public void checkIfDuplicate_test() throws Exception {
+    ArrayList<Tag> topTags = new ArrayList<Tag>();
+    topTags.add(new Tag("world", 0.5));
+    topTags.add(new Tag("hello world", 2.5));
+    topTags.add(new Tag("greetings pals", 2.0));
+    Tag next = new Tag("hello", 0.3);
+
+    assertTrue(tagsTFIDFServlet.checkIfDuplicate(topTags, next));
+  }
+
+  @Test
+  public void getTotalOccurences_test() throws Exception {
+    LinkedHashMap<String, Integer> expectedMap = new LinkedHashMap<>();
+    expectedMap.put("hello", 2);
+    expectedMap.put("my friends", 2);
+    expectedMap.put("why hello", 1);
+    expectedMap.put("nice to see", 1);
+
+    LinkedHashMap<String, Integer> occurenceMap = tagsTFIDFServlet.getTotalOccurences(groupMap);
+
+    assertEquals(occurenceMap, expectedMap);
+  }
+
+  @Test
+  public void calculateTFIDF_test() throws Exception {
+    LinkedHashMap<String, Integer> occurenceMap = tagsTFIDFServlet.getTotalOccurences(groupMap);
+
+    tagsTFIDFServlet.calculateTFIDF(groupMap, occurenceMap, mockResponse);
+
+    Key groupKey = KeyFactory.createKey("Group", GROUP_ID_1);
+    Entity groupEntity = datastore.get(groupKey);
+
+    ArrayList<Tag> tags = new ArrayList<>();
+    for (EmbeddedEntity tag : (ArrayList<EmbeddedEntity>) groupEntity.getProperty("tags")) {
+      tags.add(Tag.fromEntity(tag));
+    }
+    String tagsJson = new Gson().toJson(tags);
+
+    // expected score for "why hello" is tf * idf * weight =
+    // (1/3) * (natural log of 2) * (1 + 0.5(9/11) + 0.5(2/3.1)) = 0.40010108516...
+
+    assertTrue(tagsJson.contains("why hello") && tagsJson.contains("0.40010108516"));
```

Comment:
This is going to be a pain to maintain if you were updating how you scored TF-IDF. I think this is fine, but consider using relative rankings as a way to test rather than absolute scores, where you can prove x > y > z is maintained. This will also more meaningfully break your test when you change how things are scored relatively, you'd have to update a test which returned `and hermione` instead of `harry and hermione`, allowing the test failure/update process itself to be useful to you, the developer.